src/features/options/useGetOptions.ts

```diff
@@ -190,42 +156,124 @@ function usePreselectedOptionIndex(props: EffectProps) {
  * from a repeating group. If the options changed and the selected option (or selected row in a repeating group)
  * is gone, we should not save stale/invalid data, so we clear it.
  */
-function useRemoveStaleValues(props: EffectProps) {
+function useEffectRemoveStaleValues(props: EffectProps) {
+  const isNodeHidden = Hidden.useIsHidden(props.node);
+  const isNodesReady = NodesInternal.useIsReady();
   const [_, setForceRerender] = useState(0);
   const getAwaiting = useGetAwaitingCommits();
+  const ready = isNodesReady && !isNodeHidden;
+  const propsAsRef = useAsRef(props);
+  const readyAsRef = useAsRef(ready);
+
+  useEffect(() => {
+    function isReady() {
+      if (!readyAsRef.current) {
+        return false;
+      }
+      const awaitingCommits = getAwaiting();
+      if (awaitingCommits > 0) {
+        // We should not remove values if there are pending commits. We'll force a re-render to delay this check until
+        // the pending commits are done. This is needed because getAwaiting() is not reactive.
+        setTimeout(() => setForceRerender((r) => r + 1), 100);
+        return false;
+      }
+      return true;
+    }
+
+    if (!isReady()) {
+      return;
+    }
+
+    setTimeout(() => {
+      // If you have larger sweeping changes in the data model happening at once, such as a data processing change
+      // or a click on a CustomButton, we might not have run the hidden expressions yet when this effect runs.
+      // We'll wait a bit to make sure the hidden expressions have run before we remove the values, and if we're
+      // hidden at that point, skip the removal.
+      const { options, unsafeSelectedValues, setValues } = propsAsRef.current;
+      if (!options || !isReady()) {
+        return;
+      }
+      const itemsToRemove = unsafeSelectedValues.filter((v) => !options?.find((option) => option.value === v));
+      if (itemsToRemove.length > 0) {
+        setValues(unsafeSelectedValues.filter((v) => !itemsToRemove.includes(v)));
+      }
+    }, 200);
+  }, [getAwaiting, readyAsRef, propsAsRef, ready]);
+}
```

Comment:
I am not sure if this is a problem or not, but whenever I see setTimeouts in a use effect that are not cleared, I start thinking about what would happen if the dependencies were to change in rapid succession. I dont know how likely this is to happen though, as I dont know exactly what would cause `ready` to change.

- In the first case, would it force rerender many times? Is that necessary? or could it clear the existing timer and just postpone the rerender instead?
- If it ends up in the second timeout, I guess it would not likely call setValues several times since it is using a ref, but is it necessary to try to remove stale values several times? It only depends on the current options and selected values 🤔

Reply:
Great points! I should have cleaned up after my timeouts here. Adding that now, re-running the tests, and hoping for the best! :crossed_fingers:

Reply:
Looks good now! Marking this as resolved. :muscle: